### ISSUE TITLE
Bootstrap data by AI<>Terminal interaction

### ISSUE THREAD

**Issue by inventor2525** [LAION-AI/Open-Assistant#2287]
Basically you give a model instructions to make a simple program, and then let it interact directly with the terminal to make it and debug it. (possibly while torturing it by changing it's code to force it to test and explore more)

If you monitor test case creation (or make them manually) then you can attach the code after the test passes to the original request once it has iterated it, giving it higher quality prompt<>code response pair training data.

I've made a simple QT application "[ModelProgrammer](https://github.com/inventor2525/ModelProgrammer)" to test this concept out a little easier and it seems: Getting at least GPT3 to iterate on the code seems to take fairly few examples for it to get what it should do, but actually doing it could be a very valuable way to quickly bootstrap training data and make a model MUCH better at coding. (And make it a way more useful tool)

**Comment by olliestanley** [comment]
What exactly is this issue proposing? A future course of development for OA after we have an initial model?

(Also, potentially relevant: https://github.com/pHaeusler/micro-agent)

**Comment by inventor2525** [comment]
Yes it is post initial model. You could call it future, I'm not sure where it would fit in your timeline tbh, but as I understand it you have an initial model on hugging face, and this is inline with your vision.

@pHaeusler seems onto the same idea if that helps, but the general concept could be used to bootstrap coding related prompt > quality code & error message related prompt > fix pairs.

**Comment by andreaskoepf** [comment]
> GPT3 to iterate on the code seems to take fairly few examples for it to get what it should do

So far we only see somewhat reliable instruction following with our fine-tune of LLaMA 30B .. all other models (incl. all smaller instruct finetuned ones like Alpaca) don't seem to be capable. Although dataset quality definitely has some impact it currently seems as if nothing can really replace model-size. OpenAI models in general seem to be more usable than many other "research" models that shine on benchmarks but completely fail to deliver any real-world value (unfortunately).

**Comment by inventor2525** [comment]
Yeah, I've seen that a bit myself (I've pulled a handful on some A100 instances to play with) I actually didn't even realize you guys had tried tuning anything other than the [pythia-12b](https://huggingface.co/OpenAssistant/oasst-sft-1-pythia-12b) model yet.

I'm a little disappointed to hear the finality in your tone, I thought You.com was using a smaller model, but in fairness... I wasn't thinking of anything supper small. I mean, obviously the smaller you can make it the better it is for openness, I do wonder about how well a GPT-J or GPT-Neox-20B could learn coding if that's all it had to do, but I was thinking more like a LLaMA 60B or BLOOM.

**Comment by inventor2525** [comment]
...I wish!! I'd love to!...

I only tried out those models on some vast.ai and lambda instances as they came available. I don't have anything to get more though (I'd try this my self if I did, and the time), I've been trying to get a prototype on my last dime for a totally un-related startup, and it would need it's own compute.

Maybe in the future! -- I was wondering the feasibility of a business loan to set up or sublet cloud compute rental (4x H100s?), but not sure.
